src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,366 @@
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            breadcrumb_push <- source();
+            breadcrumb_pop  <- source();
+        }
+        Self{breadcrumb_push,breadcrumb_pop}
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Frp {
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Deref for Frp {
+    type Target = FrpInputs;
+    fn deref(&self) -> &Self::Target {
+        &self.inputs
+    }
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Frp {
+    /// Constructor.
+    pub fn new() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
+
+
+
+// ========================
+// === BreadcrumbsModel ===
+// ========================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct BreadcrumbsModel {
+    logger                : Logger,
+    display_object        : display::object::Instance,
+    pub project_name      : ProjectName,
+    breadcrumbs_container : display::object::Instance,
+    scene                 : Scene,
+    breadcrumbs           : Rc<RefCell<Vec<Breadcrumb>>>,
+    frp_outputs           : FrpOutputs,
+    current_index         : Rc<Cell<usize>>
+}
+
+impl BreadcrumbsModel {
+    /// Constructor.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S, frp:&Frp, focus_manager:&FocusManager) -> Self {
+        let scene                 = scene.into();
+        let project_name          = ProjectName::new(scene,focus_manager);
+        let logger                = Logger::new("Breadcrumbs");
+        let display_object        = display::object::Instance::new(&logger);
+        let breadcrumbs_container = display::object::Instance::new(&logger);
+        let scene                 = scene.clone_ref();
+        let breadcrumbs           = default();
+        let frp_outputs           = frp.outputs.clone_ref();
+        let current_index         = default();
+        Self{logger,display_object,scene,breadcrumbs,frp_outputs,
+            project_name,breadcrumbs_container,current_index}.init()
+    }
+
+    fn init(self) -> Self {
+        self.add_child(&self.project_name);
+        self.add_child(&self.breadcrumbs_container);
+        self.project_name.set_position(Vector3(HORIZONTAL_MARGIN,0.0,0.0));
+        self.relayout_for_project_name_width(self.project_name.width());
+        self
+    }
+
+    fn width(&self) -> f32 {
+        self.breadcrumbs.borrow().iter().map(|breadcrumb| breadcrumb.width()).sum()
+    }
+
+    fn relayout_for_project_name_width(&self, width:f32) {
+        self.breadcrumbs_container.set_position(Vector3(HORIZONTAL_MARGIN + width,0.0,0.0));
+    }
+
+    fn select_breadcrumb(&self, index:usize) -> Vec<frp::Source> {
```

Comment:
Ouch, we have a misunderstanding here. I was writing about it in my previous review - handling FRP endpoints inside model functions is hard to debug. In the previous PR you were calling frp endpoints here. Now you are returning them in vector, which IMO is as hard to track as it was earlier. What I as asking for was that model methods should not **handle** frp endpoints in any way - not calling them, nor making vectors out of them. Here,  for example, I would redefine the signature to be `fn select_breadcrumb(&self, index:usize) -> (usize,Vec<LocalCall>)` - where the `usize` is the number of pops you need to do and the resulting vec are the pushes you need to do. Then you can just handle that in frp network, without mentioning or using frp endpoints in model functions at all. This way you keep model functions in one place, and the controlling logic in other. Thi k of it this way that FRP is controller, while model i data model. You want them to be separate as much as possible.